plugins/builds/triggers/helpers.js

```diff
@@ -1063,6 +1064,38 @@ function buildsToRestartFilter(joinPipeline, groupEventBuilds, currentEvent, cur
         .filter(build => build !== null);
 }
 
+/**
+ * Check if the job is setup job with setup suffix
+ * @param  {String} jobName                 Job name
+ * @return {Boolean}
+ */
+function isStageSetup(jobName) {
+    return STAGE_SETUP_PATTERN.test(jobName);
+}
+
+/**
+ * Check if the job is a stage job
+ * @param  {String} jobName                 Job name
+ * @param  {Object} workflowGraph           Workflow Graph
+ * @return {Boolean}
+ */
+function isStageJob(workflowGraph, jobName) {
+    const jobNode = workflowGraph.nodes.find(n => n.name === jobName);
+
+    return jobNode.stageName !== undefined;
+}
+
+/**
+ * Check if the current job is a stage setup and the next job is a non-setup stage job
+ * @param {String} triggerJob                Current job
+ * @param {String} startFrom                 Next job
+ * @param {Object} workflowGraph             Workflow Graph
+ * @return {Boolean}
+ */
+function isStartFromMiddleOfStage(triggerJob, startFrom, workflowGraph) {
```

Comment:
We also need to ensure `triggerJob` and `startFrom` belongs to the same stage.
Refer https://github.com/screwdriver-cd/workflow-parser/pull/53#discussion_r1704565605